### ISSUE TITLE
onSuccess() not called

### ISSUE THREAD

**Issue by Astralgr** [serso/android-checkout#93]
Hello Sergey,

Since the Issue #78 was closed I opened a new one. I am experiencing probably same issue as user "squeeish".

I am testing my application with a static response where i try to hide a button. When I click on the button, the Payment is successful, but the onSuccess() is not triggered and the button does not hide.

Clicking a second time gives me the following error and application crashes.
`java.lang.IllegalArgumentException: Purchase flow associated with requestCode=51966 already exists`
I am calling the billing from a Fragment. Do you know why this happens ? 

However just to test it, if I add the hide function(unlock()) to the onError() and restart the app, clicking on the Button hides it so the onError() works fine.

My code:

```
private class PurchaseListener extends EmptyRequestListener<Purchase> {
    @Override
    public void onError(int response, @Nonnull Exception e) {
        super.onError(response, e);

      unlock();
    }

    @Override
    public void onSuccess(@Nonnull Purchase result) {
        Toast.makeText(getActivity(), "SUCCESS", Toast.LENGTH_SHORT).show();
         unlock();
    }
}
```

Thank You
Marios

**Comment by rpattabi** [comment]
@Astralgr In `android checkout`, there can be only one purchase flow per request code. So you run into exception when you try to create and use *one-shot* purchase flow more than once. One shot purchase flow uses a fixed request code (`0xCAFE`).

I haven't figured out the right use case for one-shot purchase flow. However, the solution to your problem is to create one instance of purchase flow (`createPurchaseFlow`) and reuse it.

**Comment by Astralgr** [comment]
@rpattabi , thank you for you repley
I would like to clarrify my problem which I should have done earlier, so sorry about that.

I have a MainActivity where inside I have a PageViewer with different Fragments. Now the content of one or more Fragments should be visible to the user only after a Purchase is made (clicking in "buy" ImageButton). As a not so experienced programmer and since this whole In-App Purchase thing is complicated, the Lirary seems a perfect way to intergrate it maybe more easily.

A good starting point is the BannerActivity of the sample code since instead of removing a banner I will remove an ImageButton. So my code is **nearly** identical to the BannerActivity.
The difference herer is that I am **inside a Fragment** and and that I cannot set the onActivityResult() and OnCreate() as "protected".

Now it seems that the onActivityResult is not called inside a Fragment. This is propably why the PurchaseFlow still exists even after I click outside of the payment box (cancel the prossecce) or even if I make a purchase successfuly

Reference:
1. There is no onActivityResult() in Fragment. For in-app Billing 
(https://stackoverflow.com/questions/26640057/there-is-no-onactivityresult-in-fragment-for-in-app-billing)
2. (https://stackoverflow.com/questions/14131171/calling-startintentsenderforresult-from-fragment-android-billing-v3)

The second link seems to have a solution but I am trying to implement it with no luck and it gets really frustrating. I would really appreciate any help here.

_Thank you and sorry for the long post but this thing really matters to me._

My Fragment code below:


```
public class FragmentThree extends Fragment implements View.OnClickListener {   

    ImageButton buy;
    View view;
    public boolean purchase_ALL = true;
    final String sku = "unlock_all";

    private ActivityCheckout mCheckout;

    public static FragmentThree newInstance(){
        FragmentThree f = new FragmentThree();
        return f;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        view = inflater.inflate(R.layout.fragment_three_layout, container, false);
        buy = (ImageButton) view.findViewById(R.id.buy);
        buy.setOnClickListener(this);

       final Billing billing = CheckoutApplication.get(getActivity()).getBilling();
        mCheckout = Checkout.forActivity(getActivity(), billing);
        mCheckout.start();
        mCheckout.loadInventory(Inventory.Request.create().loadAllPurchases(), new InventoryCallback());
        return view;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        mCheckout.onActivityResult(requestCode, resultCode, data);
        super.onActivityResult(requestCode, resultCode, data);
        }

    @Override
    public void onDestroy() {
        mCheckout.stop();
        super.onDestroy();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.buy:
                    mCheckout.startPurchaseFlow(IN_APP, sku, null, new PurchaseListener());
                    break;
                }
    }

    private class PurchaseListener extends EmptyRequestListener<Purchase> {
        @Override
        public void onError(int response, @Nonnull Exception e) {
            super.onError(response, e);
            Toast.makeText(getActivity(), "ERROR", Toast.LENGTH_SHORT).show();
            //unlock();
        }

        @Override
        public void onSuccess(@Nonnull Purchase result) {
            Log.i("buy", "Item bought");
            unlock();
        }
    }

    private void lock() {
        purchase_ALL = false;
        buy.setVisibility(View.VISIBLE);
    }

    private void unlock() {
        purchase_ALL = true;
        buy.setVisibility(View.GONE);
    }


    private class InventoryCallback implements Inventory.Callback {
        @Override
        public void onLoaded(@Nonnull Inventory.Products products) {
            final Inventory.Product product = products.get(IN_APP);
            if (!product.supported) {
                // billing is not supported, user can't purchase anything. Don't show ads in this
                // case
                return;
            }
            if (product.isPurchased(IN_APP)) {
                return;
            }
            lock();
        }
    }
}

```

**Comment by serso** [comment]
Fragment's #onActivityResult is invoked from FragmentActivity's #onActivityResult. If you don't get the call check in your activity whether #onActivityResult is called or not.

**Comment by oenukr** [comment]
Hi.

I had same issue (`onActivityResult` not called in a `Fragment`). This is because in the `PurchaseFlow` class `startIntentSenderForResult` is called on the `Activity` instance and not on the `Fragment` instance. Thus there are two options:
1. library provides something like `Checkout.forFragment(fragment, billing)` and calls `startIntentSenderForResult` on the provided fragment instance
2. developer has to get data from activity's `onActivityResult` into the `Fragment` instance, that holds reference to the `Checkout` instance

I'm using currently the second option, but would like to see the first one being implemented.

**Comment by Astralgr** [comment]
@oenukr Hi, and thank you. I am currently trying exactly what you suggest in option 2, and yes option 1 would be really great to have !

Since I don't have a FragmentActivity @serso, I am trying to do that in the MainActivity which hold the CustomViewPager with the Fragments.

Currently I can successfully get the "requestCode", "resultCode", "Intent data" in the Log of the Fragment when CANCELING (clicking outside the Purchase DialBox) the process.

> I/onActivityResult: Request CODE: 51966
> I/onActivityResult: Result Code: 0

On my "buy" Button I now have:

```
 if (resCode ==0 ){
    mCheckout.destroyPurchaseFlow();
   }
mCheckout.startPurchaseFlow(IN_APP, sku, null, new PurchaseListener());
```

**This seems to work as it cancels the Process.**

Now if I make a purchase with _sku = "android.test.purchased" i get 

> I/onActivityResult: Request CODE: 51966
> I/onActivityResult: Result Code: -1

Putting following in onSuccess() does still not work:
```
        if (resCode == RESULT_OK){
            buy.setVisibility(GONE);
        }
```

**Is this supposed to happen ?**


Just in case I can help others: Main and Fragment code below

**MainActivity** 

```
  @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {

        super.onActivityResult(requestCode, resultCode, data);
        FragmentThree fr3 = new FragmentThree();

        if (fr3 instanceof FragmentThree) {
           fr3.onActivityResult(requestCode, resultCode, data);
        }
      // setting instance of Fragment to null
      fr3 = null;
      Log.i("FR3", "Fragment: " + fr3);
    }
```

and in **Fragment** 

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        Log.i("onActivityResult", "Request CODE: " + requestCode);
        Log.i("onActivityResult", "Result Code: " + resultCode);
    }

**Comment by serso** [comment]
@oenukr thank you for your suggestion. I'll try to look into this issue during the weekend and, hopefully, prepare a fix.

**Comment by Astralgr** [comment]
@serso  thank you for considering adding a Fragment support. Should I wait for a fix then or any suggestions about my code ?
I also changed title slighly to match more the issue.

**Comment by serso** [comment]
@Astralgr you definitely should not create a new instance of `FragmentThree` in `MainActivity` but should reuse the existing one. To wait for the fix or not - it's up to you but I'll try to do it during the weekend

**Comment by serso** [comment]
Fixed in 1.0.1.
4617b00722a904feeab303bb332ff61eea6c93bd can be used as a reference implementation.
